File: src/main/java/dsl/BulkMappingRule.java
package dsl;

import java.util.Set;
import java.util.function.BiFunction;

import dsl.util.PathMatcher;
import xyz.mahmoudahmed.adapter.DataNode;

import java.util.*;

/**
 * Rule that maps multiple source fields based on a pattern.
 */
public class BulkMappingRule implements MappingRule {
    private final String sourcePattern;
    private final String targetPath;
    private final Set<String> exclusions;
    private final Set<String> inclusions;
    private final BiFunction<String, Object, Object> transformation;
    private final boolean preserveStructure;

    public BulkMappingRule(String sourcePattern, String targetPath,
                           Set<String> exclusions, Set<String> inclusions,
                           BiFunction<String, Object, Object> transformation,
                           boolean preserveStructure) {
        this.sourcePattern = sourcePattern;
        this.targetPath = targetPath;
        this.exclusions = exclusions;
        this.inclusions = inclusions;
        this.transformation = transformation;
        this.preserveStructure = preserveStructure;
    }

    @Override
    public void apply(MappingContext context, Object target) {
        if (!(target instanceof DataNode.ObjectNode)) {
            return;
        }

        DataNode sourceNode = (DataNode) context.getSourceData();
        DataNode.ObjectNode targetNode = (DataNode.ObjectNode) target;

        // Find all paths that match the pattern
        List<PathMatcher.PathMatch> matches = PathMatcher.findMatches(sourceNode, sourcePattern);

        // Apply inclusions/exclusions
        List<PathMatcher.PathMatch> filteredMatches = new ArrayList<>();
        for (PathMatcher.PathMatch match : matches) {
            String path = match.getPath();

            // Skip excluded paths
            if (isExcluded(path)) {
                continue;
            }

            // Only include paths in the inclusion list, if any
            if (!inclusions.isEmpty() && !isIncluded(path)) {
                continue;
            }

            filteredMatches.add(match);
        }

        // Process matches based on structure preservation
        if (preserveStructure) {
            applyPreservingStructure(filteredMatches, targetNode);
        } else {
            applyFlatStructure(filteredMatches, targetNode);
        }
    }

    private boolean isExcluded(String path) {
        for (String exclusion : exclusions) {
            if (path.matches(exclusion.replace("*", ".*"))) {
                return true;
            }
        }
        return false;
    }

    private boolean isIncluded(String path) {
        for (String inclusion : inclusions) {
            if (path.matches(inclusion.replace("*", ".*"))) {
                return true;
            }
        }
        return false;
    }

    private void applyPreservingStructure(List<PathMatcher.PathMatch> matches, DataNode.ObjectNode targetNode) {
        // Group matches by their path structure to maintain hierarchy
        for (PathMatcher.PathMatch match : matches) {
            String relativePath = getRelativePath(match.getPath(), sourcePattern);

            // Transform the value if needed
            Object value = match.getValue();
            if (transformation != null) {
                value = transformation.apply(match.getFieldName(), value);
            }

            // Set in the target, preserving the structure
            String fullTargetPath = targetPath;
            if (!relativePath.isEmpty()) {
                fullTargetPath = targetPath + "." + relativePath;
            }

            setValueInPath(targetNode, fullTargetPath, value);
        }
    }

    private void applyFlatStructure(List<PathMatcher.PathMatch> matches, DataNode.ObjectNode targetNode) {
        // For a flat structure, just put all fields directly under the target path
        for (PathMatcher.PathMatch match : matches) {
            // Transform the value if needed
            Object value = match.getValue();
            if (transformation != null) {
                value = transformation.apply(match.getFieldName(), value);
            }

            // Set in the target at the target path + field name
            String fieldPath = targetPath + "." + match.getFieldName();
            setValueInPath(targetNode, fieldPath, value);
        }
    }

    private String getRelativePath(String fullPath, String patternPath) {
        // Remove the base path from the full path to get the relative part
        if (patternPath.endsWith(".*")) {
            String basePath = patternPath.substring(0, patternPath.length() - 2);
            if (fullPath.startsWith(basePath)) {
                return fullPath.substring(basePath.length() + 1);
            }
        } else if (patternPath.endsWith(".**")) {
            String basePath = patternPath.substring(0, patternPath.length() - 3);
            if (fullPath.startsWith(basePath)) {
                return fullPath.substring(basePath.length() + 1);
            }
        }

        // If no match or we're dealing with a more complex pattern,
        // just return the field name
        int lastDot = fullPath.lastIndexOf('.');
        return lastDot >= 0 ? fullPath.substring(lastDot + 1) : fullPath;
    }

    private void setValueInPath(DataNode.ObjectNode targetNode, String path, Object value) {
        String[] parts = path.split("\\.");
        DataNode.ObjectNode current = targetNode;

        // Navigate to the parent node
        for (int i = 0; i < parts.length - 1; i++) {
            String part = parts[i];
            DataNode child = current.get(part);

            if (child == null || !(child instanceof DataNode.ObjectNode)) {
                DataNode.ObjectNode newNode = new DataNode.ObjectNode();
                current.getProperties().put(part, newNode);
                current = newNode;
            } else {
                current = (DataNode.ObjectNode) child;
            }
        }

        // Set the value on the final node
        String lastPart = parts[parts.length - 1];

        // Convert value to DataNode if needed
        if (value instanceof DataNode) {
            current.getProperties().put(lastPart, (DataNode) value);
        } else {
            current.getProperties().put(lastPart, convertToDataNode(value));
        }
    }

    private DataNode convertToDataNode(Object value) {
        if (value == null) {
            return DataNode.NullValue.INSTANCE;
        } else if (value instanceof String) {
            return new DataNode.StringValue((String) value);
        } else if (value instanceof Number) {
            return new DataNode.NumberValue((Number) value);
        } else if (value instanceof Boolean) {
            return new DataNode.BooleanValue((Boolean) value);
        } else if (value instanceof DataNode) {
            return (DataNode) value;
        } else if (value instanceof Map) {
            DataNode.ObjectNode objNode = new DataNode.ObjectNode();
            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
                objNode.getProperties().put(
                        entry.getKey().toString(),
                        convertToDataNode(entry.getValue())
                );
            }
            return objNode;
        } else if (value instanceof Collection) {
            DataNode.ArrayNode arrayNode = new DataNode.ArrayNode();
            for (Object item : (Collection<?>) value) {
                arrayNode.getElements().add(convertToDataNode(item));
            }
            return arrayNode;
        }

        // For other types, convert to string
        return new DataNode.StringValue(value.toString());
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/bulk/BulkMappingRule.kt
package xyz.mahmoudahmed.dsl.bulk

import xyz.mahmoudahmed.adapter.DataNode
import xyz.mahmoudahmed.dsl.util.PathMatcher
import xyz.mahmoudahmed.dsl.core.MappingContext
import xyz.mahmoudahmed.dsl.core.MappingRule

/**
 * Rule that maps multiple source fields based on a pattern with improved error handling.
 */
class BulkMappingRule(
    private val sourcePattern: String,
    private val targetPath: String,
    private val exclusions: Set<String>,
    private val inclusions: Set<String>,
    private val transformation: ((String, Any) -> Any)?,
    private val preserveStructure: Boolean
) : MappingRule {

    init {
        require(sourcePattern.isNotBlank()) { "Source pattern cannot be blank" }
        require(targetPath.isNotBlank()) { "Target path cannot be blank" }
    }

    override fun apply(context: MappingContext, target: Any) {
        val targetNode = target as? DataNode.ObjectNode ?: return
        val sourceNode = context.sourceData as? DataNode ?: return

        runCatching {
            processMatches(sourceNode, targetNode)
        }.onFailure { exception ->
            // Log error or handle according to your error strategy
            println("Error applying bulk mapping rule: ${exception.message}")
        }
    }

    private fun processMatches(sourceNode: DataNode, targetNode: DataNode.ObjectNode) {
        val matches = PathMatcher.findMatches(sourceNode, sourcePattern)
        val filteredMatches = filterMatches(matches)

        when {
            preserveStructure -> applyPreservingStructure(filteredMatches, targetNode)
            else -> applyFlatStructure(filteredMatches, targetNode)
        }
    }

    private fun filterMatches(matches: List<PathMatcher.PathMatch>): List<PathMatcher.PathMatch> {
        return matches.filter { match ->
            val path = match.path
            !isExcluded(path) && (inclusions.isEmpty() || isIncluded(path))
        }
    }

    private fun isExcluded(path: String): Boolean {
        return exclusions.any { exclusion ->
            runCatching {
                path.matches(exclusion.replace("*", ".*").toRegex())
            }.getOrDefault(false)
        }
    }

    private fun isIncluded(path: String): Boolean {
        return inclusions.any { inclusion ->
            runCatching {
                path.matches(inclusion.replace("*", ".*").toRegex())
            }.getOrDefault(false)
        }
    }

    private fun applyPreservingStructure(
        matches: List<PathMatcher.PathMatch>,
        targetNode: DataNode.ObjectNode
    ) {
        matches.forEach { match ->
            val relativePath = getRelativePath(match.path, sourcePattern)
            val transformedValue = applyTransformation(match)
            val fullTargetPath = buildTargetPath(relativePath)

            setValueInPath(targetNode, fullTargetPath, transformedValue)
        }
    }

    private fun applyFlatStructure(
        matches: List<PathMatcher.PathMatch>,
        targetNode: DataNode.ObjectNode
    ) {
        matches.forEach { match ->
            val transformedValue = applyTransformation(match)
            val fieldPath = "$targetPath.${match.fieldName}"

            setValueInPath(targetNode, fieldPath, transformedValue)
        }
    }

    private fun applyTransformation(match: PathMatcher.PathMatch): Any {
        return transformation?.let { transform ->
            runCatching {
                transform(match.fieldName, match.value)
            }.getOrElse { match.value }
        } ?: match.value
    }

    private fun buildTargetPath(relativePath: String): String {
        return if (relativePath.isEmpty()) targetPath else "$targetPath.$relativePath"
    }

    private fun getRelativePath(fullPath: String, patternPath: String): String {
        return when {
            patternPath.endsWith(".*") -> {
                val basePath = patternPath.dropLast(2)
                if (fullPath.startsWith(basePath)) {
                    fullPath.drop(basePath.length + 1)
                } else {
                    getFieldName(fullPath)
                }
            }
            patternPath.endsWith(".**") -> {
                val basePath = patternPath.dropLast(3)
                if (fullPath.startsWith(basePath)) {
                    fullPath.drop(basePath.length + 1)
                } else {
                    getFieldName(fullPath)
                }
            }
            else -> getFieldName(fullPath)
        }
    }

    private fun getFieldName(fullPath: String): String {
        return fullPath.substringAfterLast('.', fullPath)
    }

    private fun setValueInPath(targetNode: DataNode.ObjectNode, path: String, value: Any) {
        val parts = path.split(".")
        val navigationPath = parts.dropLast(1)
        val finalKey = parts.last()

        val parentNode = navigationPath.fold(targetNode) { current, part ->
            current.properties.getOrPut(part) { DataNode.ObjectNode() } as? DataNode.ObjectNode
                ?: throw IllegalStateException("Cannot navigate through non-object node at path: $part")
        }

        parentNode.properties[finalKey] = value.toDataNode()
    }

    private fun Any?.toDataNode(): DataNode = DataNodeConverter.convert(this)
}
